### PR TITLE
NFSE-4590 Cursor Read enhancements

### PR DESCRIPTION
**Pull request by gsramdasi**
Signed-off-by: Gaurav Ramdasi <10132364+gsramdasi@users.noreply.github.com>

## Description
The cursor read api can now return just keys if the valur parameter is set to NULL. There's a new api called hse_kvs_cursor_read_copy(...) which allows the user to pass in buffers for the key and value and hse will copy the key-value pait into these buffers. This api will also return just the key if the value buffer is set to NULL.

## Issue(s) Addressed
NFSE-4590

### CODE REVIEW

**Comment by tristan957** [comment]
I would personally like to see this as two different PRs. Otherwise the commits will get squashed together

**Comment by alexttx** [comment]
Update `tests/functional/api/cursor_api_test.c` to exercise the new API?

**Comment by gsramdasi** [comment]
> I would personally like to see this as two different PRs. Otherwise the commits will get squashed together

Right. But the changes needed for these two tasks have a lot in common, so it made sense to club them together.

File: lib/binding/kvdb_interface.c [hunk 1]
```diff
@@ -1060,13 +1060,17 @@ hse_kvs_cursor_read(
...
+    size_t vlen_dummy;
 
-    if (HSE_UNLIKELY(!cursor || !key || !klen || !val || !vlen || !eof || flags != 0))
+    if (HSE_UNLIKELY(!cursor || !key || !klen || !eof || flags != 0))
         return merr(EINVAL);
 
     err = ikvdb_kvs_cursor_read(cursor, flags, key, klen, val, vlen, eof);
     ev(err);
 
+    if (!vlen)
```
Comment (tristan957) [inline lib/binding/kvdb_interface.c:1071]: Why is this necessary?

File: lib/binding/kvdb_interface.c [hunk 2]
```diff
@@ -1060,13 +1060,17 @@ hse_kvs_cursor_read(
     bool *                 eof)
 {
     merr_t err;
+    size_t vlen_dummy;
 
-    if (HSE_UNLIKELY(!cursor || !key || !klen || !val || !vlen || !eof || flags != 0))
+    if (HSE_UNLIKELY(!cursor || !key || !klen || !eof || flags != 0))
```
Comment (tristan957) [inline lib/binding/kvdb_interface.c:1065]: Is klen really required? If user doesn't want that information, shouldn't we support it?

File: lib/kvdb/ikvdb.c [hunk 3]
```diff
@@ -2595,17 +2594,68 @@ ikvdb_kvs_cursor_read(
...
     if (*eof)
         return 0;
 
-    *key = kvt.kvt_key.kt_data;
-    *key_len = kvt.kvt_key.kt_len;
+    *key = kvs_cursor_key_copy(cur, NULL, 0, key_len);
+    if (val) {
+        *val = kvs_cursor_val_copy(cur, NULL, 0, val_len);
+        if (ev(!val))
+            return merr(EBUG);
```
Comment (tristan957) [inline lib/kvdb/ikvdb.c:2607]: Having a hard time understanding the EBUG usage here. It looks like one of the fallible operations in val_copy() is the decompress operation. Why not continue to carry the original error forward instead of masking it? And then there is a comparison with a min_t as well. To me it seems like EBUG is a misnomer, and anywhere we have EBUG should be an assert or an abort.

**Comment by tristan957** [comment]
Make sure you are rebased on master

**Review by nabeelmmd — APPROVED** [review]

File: lib/binding/kvdb_interface.c (hunk 1 above)
Comment (gsramdasi) [reply lib/binding/kvdb_interface.c:1071]: I had added it for the perf counter update below, but I realize now that that is incorrect when we don't read the values. Fixed it.

File: lib/binding/kvdb_interface.c (hunk 2 above)
Comment (gsramdasi) [reply lib/binding/kvdb_interface.c:1065]: Well, just having a pointer for the key without knowing its length doesn't seem meaningful.

File: lib/kvdb/ikvdb.c (hunk 3 above)
Comment (gsramdasi) [reply lib/kvdb/ikvdb.c:2607]: Yes. Fixed it.

File: lib/binding/kvdb_interface.c (hunk 2 above)
Comment (alexttx) [reply lib/binding/kvdb_interface.c:1143]: Keys and values are not necessarily NULL terminated.  So the length is always explicit.

File: include/hse/hse.h [hunk 4]
```diff
@@ -863,6 +863,50 @@ hse_kvs_cursor_read(
...
+hse_kvs_cursor_read_copy(
+    struct hse_kvs_cursor *cursor,
+    const unsigned int     flags,
+    const void *           keybuf,
+    size_t                 keybuf_sz,
+    size_t *               key_len,
+    const void *           valbuf,
+    size_t                 valbuf_sz,
+    size_t *               val_len,
+    bool *                 eof);
```
Comment (alexttx) [inline include/hse/hse.h:908]: Need to make `keybuf` and `valbuf` non const.

File: lib/binding/kvdb_interface.c [hunk 5]
```diff
@@ -1140,20 +1140,67 @@ hse_kvs_cursor_read(
 {
     merr_t err;
 
-    if (HSE_UNLIKELY(!cursor || !key || !klen || !val || !vlen || !eof || flags != 0))
+    if (HSE_UNLIKELY(!cursor || !key || !klen || !eof || flags != 0))
+        return merr(EINVAL);
+
+    if (HSE_UNLIKELY(!!val ^ !!vlen))
         return merr(EINVAL);
```
Comment (alexttx) [inline lib/binding/kvdb_interface.c:1147]: We could be less restrictive and define semantics as follows:
- read value if and only if `val != NULL and vlen != NULL and *vlen > 0`
- it is not an error if val or vlen are NULL
Thoughts?

File: lib/kvs/kvs_cursor.c [hunk 6]
```diff
@@ -1100,32 +1100,63 @@ ikvs_cursor_seek(struct kvs_cursor_impl *cursor, const void *key, size_t klen)
...
+    struct kvs_cursor_impl *cur = cursor_h2r(cursor);
+    void                   *buf = (void *)kbuf;
+    size_t                  bufsz = kbufsz;
+    uint                    keylen;
+    const void             *key;
+
+    if (!buf) {
+        buf = (void *)cur->kci_buf;
+        bufsz = HSE_KVS_KEY_LEN_MAX;
+    }
```
Comment (alexttx) [inline lib/kvs/kvs_cursor.c:1115]: Need to make the read_copy callstack not use const key and value buffers.  And then drop these casts that drop the const qualifier.

File: lib/kvs/kvs_cursor.c [hunk 7]
```diff
@@ -1100,32 +1100,63 @@ ikvs_cursor_seek(struct kvs_cursor_impl *cursor, const void *key, size_t klen)
...
+    uint               clen = cur->kci_elem_last.kce_complen;
+    void              *buf = (void *)vbuf;
+    size_t             bufsz = vbufsz;
     merr_t             err = 0;
 
-    kvt->kvt_key.kt_data = key_obj_copy(
-        cursor->kci_buf, HSE_KVS_KEY_LEN_MAX, (uint *)&kvt->kvt_key.kt_len, cursor->kci_last);
+    if (!buf) {
+        buf = cur->kci_buf + HSE_KVS_KEY_LEN_MAX;
+        bufsz = HSE_KVS_VALUE_LEN_MAX;
```
Comment (alexttx) [inline lib/kvs/kvs_cursor.c:1137]: `kci_buf` is a flexible array at the end of `struct kvs_cursor_impl` and is always `HSE_KVS_KEY_LEN_MAX + HSE_KVS_VALUE_LEN_MAX` bytes in size.  Since the size is fixed, I don't see any benefit to using a flexible array.  The code would be simplified if `kci_key[HSE_KVS_KEY_LEN_MAX]` and `kci_val[HSE_KVS_VALUE_LEN_MAX ]` were defined in the struct instead of the flexible array.

**Comment by beckerg** [comment]
That code has changed quite a bit, I suspect there was a reason for it to be flexible at one point, but it looks fixed-size now.  The pointers into the buffer might need to be variable rather than named arrays, dunno though, you'd have to audit their usage to find out...

File: lib/binding/kvdb_interface.c (hunk 2 above)
Comment (tristan957) [reply lib/binding/kvdb_interface.c:1143]: I understand, but that just seems like either a failure from the user perspective or I know the length of my keys beforehand. If HSE doesn't require the klen to be passed, then I don't understand why it is an error.

File: lib/kvs/kvs_cursor.c (hunk 7 above)
Comment (tristan957) [reply lib/kvs/kvs_cursor.c:1137]: Agree with you

File: include/hse/hse.h [hunk 8]
```diff
@@ -863,6 +863,50 @@ hse_kvs_cursor_read(
...
+ * @remark @p val must not be NULL.
+ * @remark @p val_len must not be NULL.
+ * @remark @p eof must not be NULL.
+ *
+ * @returns Error status
+ */
+hse_err_t
+hse_kvs_cursor_read_copy(
+    struct hse_kvs_cursor *cursor,
+    const unsigned int     flags,
```
Comment (tristan957) [inline include/hse/hse.h:901]: Take the const away from the flags argument, but keep it in the definition.

File: include/hse/hse.h [hunk 9]
```diff
@@ -863,6 +863,50 @@ hse_kvs_cursor_read(
...
+ * @param[in,out] valbuf: Buffer into which the next key's value will be copied.
+ * @param valbuf_sz: Size of @p valbuf
+ * @param[out] val_len: Length of @p val.
+ * @param[out] eof: If true, no more key-value pairs in sequence.
+ *
+ * @remark @p cursor must not be NULL.
+ * @remark @p key must not be NULL.
+ * @remark @p key_len must not be NULL.
+ * @remark @p val must not be NULL.
+ * @remark @p val_len must not be NULL.
```
Comment (tristan957) [inline include/hse/hse.h:893]: I don't think this remark is valid?

File: lib/binding/kvdb_interface.c [hunk 10]
```diff
@@ -1140,20 +1140,67 @@ hse_kvs_cursor_read(
...
+}
+
+hse_err_t
+hse_kvs_cursor_read_copy(
+    struct hse_kvs_cursor *cursor,
+    const unsigned int     flags,
+    const void *           keybuf,
+    size_t                 keybuf_sz,
+    size_t *               key_len,
+    const void *           valbuf,
```
Comment (tristan957) [inline lib/binding/kvdb_interface.c:1171]: I always make thes {type} *const {var}, but I understand it is controversial :)

File: lib/binding/kvdb_interface.c [hunk 11]
```diff
@@ -1140,20 +1140,67 @@ hse_kvs_cursor_read(
...
+        return merr(EINVAL);
+
+    if (HSE_UNLIKELY(!valbuf && valbuf_sz > 0))
+        return merr(EINVAL);
+
+    if (HSE_UNLIKELY(!!valbuf ^ !!val_len))
+        return merr(EINVAL);
+
+    err = ikvdb_kvs_cursor_read_copy(cursor, flags, keybuf, keybuf_sz, key_len,
+                                valbuf, valbuf_sz, val_len, eof);
```
Comment (tristan957) [inline lib/binding/kvdb_interface.c:1188]: Positioning of second line is funky. Many camps one could be in. I sometimes find myself in the camp of double indent the second line.

File: lib/include/hse_ikvdb/ikvdb.h [hunk 12]
```diff
@@ -501,6 +501,18 @@ ikvdb_kvs_cursor_read(
     size_t *               val_len,
     bool *                 eof);
 
+merr_t
+ikvdb_kvs_cursor_read_copy(
+    struct hse_kvs_cursor *cur,
+    const unsigned int     flags,
+    const void *           keybuf,
```
Comment (tristan957) [inline lib/include/hse_ikvdb/ikvdb.h:508]: Note to remove const here as well.

File: lib/include/hse_ikvdb/kvs.h [hunk 13]
```diff
@@ -125,7 +125,13 @@ kvs_cursor_seek(
     struct kvs_ktuple *    kt);
 
 merr_t
-kvs_cursor_read(struct hse_kvs_cursor *cursor, struct kvs_kvtuple *kvt, bool *eof);
+kvs_cursor_read(struct hse_kvs_cursor *cursor, unsigned int flags, bool *eof);
+
+void
+kvs_cursor_key_copy(struct hse_kvs_cursor *cursor, const void *kbuf, size_t kbufsz, const void **key_out, size_t *klen_out);
```
Comment (tristan957) [inline lib/include/hse_ikvdb/kvs.h:131]: Probably remove const here.

File: lib/kvs/kvs_cursor.c [hunk 14]
```diff
@@ -1100,32 +1100,63 @@ ikvs_cursor_seek(struct kvs_cursor_impl *cursor, const void *key, size_t klen)
...
         if (ev(err))
             return err;
 
-        if (ev(outlen != vt->vt_xlen))
-            return merr(EBUG);
+        if (ev(outlen != min_t(u64, vt->vt_xlen, bufsz)))
+            return EBUG;
     } else {
-        memcpy(vbuf, vt->vt_data, vt->vt_xlen);
+        memcpy(buf, vt->vt_data, min_t(u64, vt->vt_xlen, bufsz));
```
Comment (tristan957) [inline lib/kvs/kvs_cursor.c:1153]: Use uint64_t

File: tests/unit/kvs/kvs_cursor_test.c [hunk 15]
```diff
@@ -93,100 +93,97 @@ insert_key_multiple(struct mtf_test_info *lcl_ti, char *pfx, uint nkeys)
...
 void
-expect_key(struct mtf_test_info *lcl_ti, struct kvs_ktuple *kt_found, char *expected)
+expect_pfx(struct mtf_test_info *lcl_ti, const void *key, size_t klen, const char *pfx)
 {
     int rc;
 
-    rc = strcmp(kt_found->kt_data, expected);
+    ASSERT_TRUE(strlen(pfx) <= klen);
+
+    rc = strncmp(pfx, key, strlen(pfx));
```
Comment (tristan957) [inline tests/unit/kvs/kvs_cursor_test.c:113]: Maybe use strnlen since you know the max size of a prefix from hse/limits.h

**Review by tristan957 — APPROVED** [review]

File: lib/binding/kvdb_interface.c (hunk 5 above)
Comment (tristan957) [reply lib/binding/kvdb_interface.c:1147]: Still in the camp of vlen doesn't matter if HSE doesn't require it!

File: tests/unit/kvs/kvs_cursor_test.c (hunk 15 above)
Comment (alexttx) [reply tests/unit/kvs/kvs_cursor_test.c:113]: I don't see value in `strnlen` when the input string is properly terminated.

File: include/hse/hse.h (hunk 9 above)
Comment (alexttx) [reply include/hse/hse.h:893]: You are correct.  I will edit.

**Comment by beckerg** [comment]
Sorry, I meant "shouldn't be using", but I see this is a test which is probably only using strings, so nevermind...

File: lib/kvs/kvs_cursor.c (hunk 14 above)
Comment (alexttx) [reply lib/kvs/kvs_cursor.c:1153]: This function (kvs_cursor_val_copy) seems like it misuses `vt_xlen`.  When a value is compressed, doesn't `vt_xlen == (clen<<32)|vlen`?

**Review by beckerg — CHANGES_REQUESTED** [review]

File: lib/binding/kvdb_interface.c (hunk 2 above)
Comment (alexttx) [reply lib/binding/kvdb_interface.c:1143]: Will fix this.

File: lib/binding/kvdb_interface.c (hunk 11 above)
Comment (tristan957) [reply lib/binding/kvdb_interface.c:1188]: Always feels icky to me. Inevitably you end up at the column limit and you end up with one parameter per line

Comment (beckerg) [reply lib/binding/kvdb_interface.c:1188]: Well, frankly I try to avoid the situation by using shorter function names and fewer arguments.  But when it can't be helped I  use whatever fits into the fewest number of lines that looks nice.  Ideally if I grep for the function name I'll see all the args, which is why I like to see at least some args, but if all the args will fit on the second line then I might do that, in which case the indent is just a regularly spaced indent (which can clash with a following open brace).
But I don't follow how you get to "one parameter per line" ?

File: lib/binding/kvdb_interface.c (hunk 2 above)
Comment (alexttx) [reply lib/binding/kvdb_interface.c:1143]: Will fix this in a different PR.

File: lib/binding/kvdb_interface.c (hunk 5 above)
Comment (alexttx) [reply lib/binding/kvdb_interface.c:1147]: Will fix in another PR

File: lib/binding/kvdb_interface.c (hunk 11 above)
Comment (tristan957) [reply lib/binding/kvdb_interface.c:1188]:
```
// nested deeply running up to column limit 100
my_func(p1,
              p2,
              p3);

File: lib/kvs/kvs_cursor.c (hunk 14 above)
Comment (alexttx) [reply lib/kvs/kvs_cursor.c:1159]: The reason this works is because vtuples that come out of cursor merge loops use vt_xlen differently.  Not going to address that as part of this PR.

File: lib/binding/kvdb_interface.c (hunk 11 above)
Comment (beckerg) [reply lib/binding/kvdb_interface.c:1188]: Why would you torture future readers of the code like that?

File: lib/binding/kvdb_interface.c [hunk 16]
```diff
@@ -1060,13 +1060,17 @@ hse_kvs_cursor_read(
...
 
-    if (HSE_UNLIKELY(!cursor || !key || !klen || !val || !vlen || !eof || flags != 0))
+    if (HSE_UNLIKELY(!cursor || !key || !klen || !eof || flags != 0))
         return merr(EINVAL);
 
     err = ikvdb_kvs_cursor_read(cursor, flags, key, klen, val, vlen, eof);
     ev(err);
 
+    if (!vlen)
+        vlen = &vlen_dummy;
```
[the remark this answers is not in the document]
Comment (alexttx) [reply lib/binding/kvdb_interface.c]: Fixed.

File: lib/kvdb/ikvdb.c [hunk 17]
```diff
@@ -2595,17 +2594,68 @@ ikvdb_kvs_cursor_read(
...
+    }
```
[the remark this answers is not in the document]
Comment (alexttx) [reply lib/kvdb/ikvdb.c:2609]: Fixed.

File: lib/kvdb/ikvdb.c [hunk 18]
```diff
@@ -2595,17 +2594,68 @@ ikvdb_kvs_cursor_read(
...
+
+    perfc_lat_record(
+        cur->kc_pkvsl_pc,
+        cur->kc_flags & HSE_CURSOR_CREATE_REV ? PERFC_LT_PKVSL_KVS_CURSOR_READREV
+                                                : PERFC_LT_PKVSL_KVS_CURSOR_READFWD,
+        tstart);
```
[the remark this answers is not in the document]
Comment (alexttx) [reply lib/kvdb/ikvdb.c:2615]: Future work.

File: lib/kvs/kvs_cursor.c [hunk 19]
```diff
@@ -1100,32 +1100,58 @@ ikvs_cursor_seek(struct kvs_cursor_impl *cursor, const void *key, size_t klen)
...
+    if (!buf) {
+        buf = (void *)cur->kci_buf;
+        bufsz = HSE_KVS_KEY_LEN_MAX;
+    }
+
+    key = key_obj_copy(buf, bufsz, &keylen, cur->kci_last);
+    *klen = keylen;
+
+    return key;
+}
```
[the remark this answers is not in the document]
Comment (alexttx) [reply lib/kvs/kvs_cursor.c:1127]: Fixed.

File: lib/kvs/kvs_cursor.c [hunk 20]
```diff
@@ -1100,32 +1100,58 @@ ikvs_cursor_seek(struct kvs_cursor_impl *cursor, const void *key, size_t klen)
...
+        if (ev(outlen != min_t(u64, vt->vt_xlen, bufsz)))
+            return NULL;
     } else {
-        memcpy(vbuf, vt->vt_data, vt->vt_xlen);
+        memcpy(buf, vt->vt_data, min_t(u64, vt->vt_xlen, bufsz));
     }
 
-    kvs_vtuple_init(&kvt->kvt_value, vbuf, vt->vt_xlen);
-    return err;
+    return buf;
```
[the remark this answers is not in the document]
Comment (alexttx) [reply lib/kvs/kvs_cursor.c]: Not a bug since cursors do not pack data into the upper 32-bits of vt_xlen.  I don't like the fact that vt_xlen is used in different ways depending on context, but it's not a bug that needs to be fixed in this PR.

File: lib/kvs/kvs_cursor.c [hunk 21]
```diff
@@ -1136,22 +1137,26 @@ kvs_cursor_val_copy(struct hse_kvs_cursor *cursor, const void *vbuf, size_t vbuf
...
+            return EBUG;
     } else {
         memcpy(buf, vt->vt_data, min_t(u64, vt->vt_xlen, bufsz));
     }
 
-    return buf;
+    if (val_out)
+        *val_out = buf;
+
+    return 0;
```
[the remark this answers is not in the document]
Comment (alexttx) [reply lib/kvs/kvs_cursor.c:1168]: See related comments in this PR.

File: lib/kvs/kvs_cursor.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (beckerg) [reply lib/kvs/kvs_cursor.c]: Ok, so future we we make it a union with a different name for use by cursors...  Or some such, too confusing as-is...

File: include/hse/hse.h [hunk 22]
```diff
@@ -985,6 +985,50 @@ hse_kvs_cursor_read(
...
 
+/** @brief Iteratively access the elements pointed to by the cursor.
+ *
+ * Read a key-value pair from the cursor, advancing the cursor past its current
+ * location. The key-value pair will be copied into the user's buffer(s). If the
+ * argument @p valbuf is NULL, only the key is read.
+ *
+ * @note If the cursor is at EOF, attempts to read from it will not change the
+ * state of the cursor.
+ * @note This function is thread safe across disparate cursors.
```
[the remark this answers is not in the document]
Comment (alexttx) [reply include/hse/hse.h:874]: Comment is not incorrect though is it?  It is thread-safe if the same cursor object isn't used concurrently in different threads.

File: include/hse/hse.h (hunk 8 above)
Comment (alexttx) [reply include/hse/hse.h]: Gone from both.

File: include/hse/hse.h (hunk 9 above)
Comment (alexttx) [reply include/hse/hse.h:893]: Or not.  Will revisit making ptrs optional.

File: include/hse/hse.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (beckerg) [reply include/hse/hse.h:874]: Saying that something is thread-safe implies multiple threads can access the object safely, and trying to qualify that with some verbiage about just one object is very confusing.  So nothing here is really thread-safe, it's all single-threaded (with the caveat that you can access the same object from different threads from completely non-overlapping calls...  I think.  But that's no thread-safe)

It'll be thread-safe once we put in the requisite locking like we did for ctxn.

File: lib/binding/kvdb_interface.c (hunk 11 above)
Comment (tristan957) [reply lib/binding/kvdb_interface.c:1188]: That is the flaw I am talking about with lining up parameters under each other.

Comment (beckerg) [reply lib/binding/kvdb_interface.c:1188]: Yeah, to be clear, I meant indent the second line starting under the first arg, which was not mean to imply line them each up one by one under the first arg (which is just awful unless you're getting paid by LOC and won't be around to maintain it).

File: include/hse/hse.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (alexttx) [reply include/hse/hse.h:874]: Will change to "cursor objects are not thread safe"

**Review by beckerg — APPROVED** [review]
